Changes for early fetching of methods

ricea opened the pull request:
Stream classes are changing to get methods from the underlying object
during construction rather than when the methods are called.

Change expectations of when method lookup happens.

See https://github.com/whatwg/streams/issues/691 for background.

<!-- Reviewable:start -->

<!-- Reviewable:end -->

ricea committed:
Add properties tests for TransformStream

Add tests to verify that TransformStream and
TransformStreamDefaultController have the correct properties, and that
those properties have the correct properties.

Also confirm that transformer objects are called in the correct ways,
and not called in unexpected ways.

ricea committed:
Changes for early fetching of methods

Stream classes are changing to get methods from the underlying object
during construction rather than when the methods are called.

Change expectations of when method lookup happens.

See https://github.com/whatwg/streams/issues/691 for background.

w3c-bots commented:
# Build PENDING

Started: None
Finished: None





View more information about this build on:

- [WPT PR Status](http://pulls.web-platform-tests.org/build/21290)
- [TravisCI](https://travis-ci.org/w3c/web-platform-tests/builds/309549978)

domenic commented on streams/readable-streams/bad-underlying-sources.js:
```diff
@@ -82,9 +80,11 @@ promise_test(t => {
+    })
   ]);
 
 }, 'Underlying source pull: throwing getter (second pull)');
```
This test might benefit from a comment saying that the test is that it's only gotten once. Indeed, asserting that counter === 1 after the last assert_object_equals might be good.

ricea replied:
Done.

domenic commented on streams/writable-streams/properties.js:
```diff
@@ -194,28 +194,32 @@ for (const method in sinkMethods) {
+      assert_array_equals(touchedProperties, ['type', 'write', 'close', 'abort', 'start'],
+                          'no properties should be accessed on method call');
+    });
+  }, `unexpected properties should not be accessed when calling sink method ${method}`);
```
Should we clone this test for readable too, while we're here?

(What a great test, BTW...)

ricea replied:
I've been thinking about throwing away readable-stream's properties.js and use one based on this instead. But I'd have to check carefully that I wasn't losing any coverage, which is why I haven't done it yet.

domenic reviewed (approved):
Some suggestions, but overall looks good. I think we should add transform stream tests too though, probably before merging.

ricea committed:
Clarify 'Underlying source pull: throwing getter (second pull)' test

This test now verifies that the getter is only called once, so make that
explicit.

ricea committed:
Add TransformStream properties.js test

Add a properties.js test for TransformStream based on the one for
WritableStream. This verifies that the correct properties are accessed in the
right order during TransformStream construction, and that no properties are
accessed afterwards.

ricea commented:
Corresponding standard change is https://github.com/whatwg/streams/pull/860 (still WIP at this time).

domenic reviewed (approved):
LGTM!
